contrib/golang/upstreams/http/tcp/source/upstream_request.cc

```diff
@@ -0,0 +1,419 @@
+  if (req_->configId == 0) {
+    req_->setWeakFilter(weak_from_this());
+    req_->configId = config_->getConfigId();
+    return true;
+  }
+  return false;
+}
+
+void TcpUpstream::encodeData(Buffer::Instance& data, bool end_stream) {
+  ENVOY_LOG(debug, "tcp upstream encodeData: {}", data.toString());
+
+  ProcessorState& state = encoding_state_;
+  Buffer::Instance& buffer = state.doDataList.push(data);
+  auto s = dynamic_cast<processState*>(&state);
+  state.setFilterState(FilterState::ProcessingData);
+
+  GoUint64 if_end_stream = dynamic_lib_->envoyGoEncodeData(
+    s, end_stream ? 1 : 0, reinterpret_cast<uint64_t>(&buffer), buffer.length());
+  if (if_end_stream == static_cast<GoUint64>(EndStreamType::NotEndStream)) {
+    end_stream = false;
+  } else {
+    end_stream = true;
+  }
+
+  state.setFilterState(FilterState::Done);
+  state.doDataList.moveOut(data);
+
+  upstream_conn_data_->connection().write(data, end_stream);
+}
+
+void TcpUpstream::doHeadersGo(ProcessorState& state, const Envoy::Http::RequestOrResponseHeaderMap& headers, bool end_stream) {
+  state.headers = &headers;
+  auto s = dynamic_cast<processState*>(&state);
+  state.setFilterState(FilterState::ProcessingData);
+
+  dynamic_lib_->envoyGoEncodeHeader(
+  s, end_stream ? 1 : 0, headers.size(),headers.byteSize());
+
+  }
+
+Envoy::Http::Status TcpUpstream::encodeHeaders(const Envoy::Http::RequestHeaderMap& headers,
+                                               bool end_stream) {
+   
+  ENVOY_LOG(debug, "tcp upstream encodeHeaders: {}", headers);
+
+  initRequest();
+
+  ProcessorState& state = encoding_state_;
+  doHeadersGo(state, headers, end_stream);
+  
+  // Headers should only happen once, so use this opportunity to add the proxy
+  // proto header, if configured.
+
+  ASSERT(route_entry_ != nullptr);
+  if (route_entry_->connectConfig().has_value()) {
+    Buffer::OwnedImpl data;
+    const auto& connect_config = route_entry_->connectConfig();
+    if (connect_config->has_proxy_protocol_config() &&
+        upstream_request_->connection().has_value()) {
+      Extensions::Common::ProxyProtocol::generateProxyProtoHeader(
+          connect_config->proxy_protocol_config(), *upstream_request_->connection(), data);
+    }
+
+    if (data.length() != 0 || end_stream) {
+      upstream_conn_data_->connection().write(data, end_stream);
+    }
+  }
+
+  // TcpUpstream::encodeHeaders is called after the UpstreamRequest is fully initialized. Alsoc use
+  // this time to synthesize the 200 response headers downstream to complete the CONNECT handshake.
+  Envoy::Http::ResponseHeaderMapPtr headersToDownstream{
+      Envoy::Http::createHeaderMap<Envoy::Http::ResponseHeaderMapImpl>(
+          {{Envoy::Http::Headers::get().Status, "200"}})};
+  upstream_request_->decodeHeaders(std::move(headersToDownstream), false);
+  return Envoy::Http::okStatus();
+}
+
+void TcpUpstream::encodeTrailers(const Envoy::Http::RequestTrailerMap&) {
+  Buffer::OwnedImpl data;
```

Comment:
ditto, we'd better golang convert trailers to data buffer, but we can just add a TODO for it now, it's not a common use case.

Reply:
got it, will finish it in the near future.